xarray/core/dataarray.py

```diff
@@ -462,23 +460,40 @@ def _replace_maybe_drop_dims(
             )
         return self._replace(variable, coords, name, indexes=indexes)
 
-    def _overwrite_indexes(self, indexes: Mapping[Hashable, Any]) -> "DataArray":
-        if not len(indexes):
+    def _overwrite_indexes(
+        self,
+        indexes: Mapping[Hashable, Index],
+        coords: Optional[Mapping[Hashable, Variable]] = None,
```

Comment:
```suggestion
        coords: Optional[Mapping[Any, Variable]] = None,
```
A small note here but using hashable doesn't seem to be such a good idea. See #5690.
And if you do drive-by fixes to tests, add some typing to them so mypy can check them as well.

Reply:
Ah I see, thanks! I didn't pay attention enough to #5690 (I'm still not super familiar with all Python annotation subtleties to be honest).

Reply:
> if you do drive-by fixes to tests, add some typing to them so mypy can check them as well.

I suggest doing this in follow-up PRs (except for new tests) as this one is already getting big and messy.

Reply:
Sure, all I'm saying that if you're looking at or editing a test add at least a quick `-> None` so mypy checks it.